pruner/state.go

```diff
@@ -0,0 +1,100 @@
+package pruner
+
+import (
+	"fmt"
+	"sync"
+
+	quamina "quamina/lib"
+)
+
+// LivePatternsState represents the required capabilities for
+// maintaining the set of live patterns.
+type LivePatternsState interface {
+	// Add adds a new pattern or updates an old pattern.
+	//
+	// Note that multiple patterns can be associated with the same
+	// X.
+	Add(x quamina.X, pattern string) error
+
+	Delete(x quamina.X) (bool, error)
+
+	// Iterate calls the given function for every stored pattern.
+	Iterate(func(x quamina.X, pattern string) error) error
+
+	// Contains returns true if x is in the live set; false
+	// otherwise.
+	//
+	// Since a pattern can't be the empty string, that zero value
+	// indicates no corresponding pattern.
+	Contains(x quamina.X) (bool, error)
+}
+
+type (
+	stringSet map[string]nothing
+	nothing   struct{}
+)
+
+var na = nothing{}
+
+// MemState is a LivePatternsState that is just a map (with a
+// RWMutex).
+//
+// Since the LivePatternsState implementation can be provided to the
+// application, we're keeping things simple here initially.
+type MemState struct {
+	lock sync.RWMutex
+	m    map[quamina.X]stringSet
+}
+
+func NewMemState() *MemState {
+	// Accept initial size as a parameter?
+	return &MemState{
+		m: make(map[quamina.X]stringSet),
+	}
+}
+
+var ErrExists = fmt.Errorf("pattern already exists for that X")
+
+func (s *MemState) Add(x quamina.X, pattern string) error {
+	s.lock.Lock()
+	ps, have := s.m[x]
+	if !have {
+		ps = make(stringSet)
+		s.m[x] = ps
+	}
+	ps[pattern] = na
+	s.lock.Unlock()
+	return nil
+}
+
+func (s *MemState) Contains(x quamina.X) (bool, error) {
+	s.lock.RLock()
+	_, have := s.m[x]
+	s.lock.RUnlock()
+	return have, nil
+}
+
+func (s *MemState) Delete(x quamina.X) (bool, error) {
+	s.lock.Lock()
+	_, had := s.m[x]
+	if had {
+		delete(s.m, x)
```

Comment:
To keep an accurate count of the number of live patterns.

But I just realized that code is wrong because of the one-to-many ids-to-patterns capability!

Will fix.

Reply:
PrunerMatcher.DeletePattern() doesn't need to return (bool,error) and could just return the error. Better to do it that way? Sometimes I like to know if my Delete found what I think it should find. Not a big deal.

Reply:
See c4d3b4d for at least one fix. Happy to drop the bool from DeletePattern.

Reply:
Yeah, if you don't mind I'd prefer to drop that boolean.  In fact, it's not obvious to me that DeletePattern has any useful return value at all.  How can an error occur?

But, you implicitly raise a good point.  Is there a case for an API to discover what patterns are current in a Matcher?  Probably not just yet, but let's keep it in mind.

Reply:
Oh wait, DeletePattern(nil) is probably a bug, so I guess we should retain the error.